lib/internal/Magento/Framework/Cache/Frontend/Decorator/TagScope.php

```diff
@@ -61,10 +61,8 @@ public function clean($mode = \Zend_Cache::CLEANING_MODE_ALL, array $tags = [])
     {
         if ($mode == \Zend_Cache::CLEANING_MODE_MATCHING_ANY_TAG) {
             $result = false;
-            foreach ($tags as $tag) {
-                if (parent::clean(\Zend_Cache::CLEANING_MODE_MATCHING_TAG, [$tag, $this->getTag()])) {
-                    $result = true;
-                }
+            if (parent::clean(\Zend_Cache::CLEANING_MODE_MATCHING_ANY_TAG, array_merge($tags, [$this->getTag()]))) {
```

Comment:
@raivisdejus I don't believe it will work correctly, because we need to fluch cache for each of provided tags that are going together with MAGE (value of $this->getTag()), like (tag1  AND MAGE) OR (tag2 AND MAGE).

After your change seems like we'll have something like this: tag1 OR tag2 OR MAGE.

Reply:
@ihor-sviziev thanks for quick reply, valid points.

Why do we need the `tag1` and `MAGE` combination? 
If product `123` changed it should be sufficient to flush tag `cat_p_123`

As you mentioned I should remove the `MAGE` tag, but flushing a list of tags like `cat_p_123` should do what is intended.

Reply:
As I undertood - in case if we'll use the same cache storage between few applications, in this case we need to make sure we're flushing only cache entities that tagger with MAGE (magento2). Also I think it's possible to have different tags between different magento installations.

Reply:
Thanks, Will see if I can refactor this to be good for all cases.